File: day05/main.exs
upperHalf = fn c ->
  {min, max} = c
  range = max - min
  newRange = floor(range / 2)
  {max - newRange, max}
end

lowerHalf = fn c ->
  {min, max} = c
  range = max - min
  newRange = floor(range / 2)
  {min, min + newRange}
end

calc = fn i ->
  {rows, seats} = String.split_at(i, 7)

  {row, _} =
    Enum.reduce(String.graphemes(rows), {0, 127}, fn s, acc ->
      case s do
        "F" -> lowerHalf.(acc)
        "B" -> upperHalf.(acc)
      end
    end)

  {seat, _} =
    Enum.reduce(String.graphemes(seats), {0, 7}, fn s, acc ->
      case s do
        "L" -> lowerHalf.(acc)
        "R" -> upperHalf.(acc)
      end
    end)

  row * 8 + seat
end

list =
  File.read!("input.txt")
  |> String.trim()
  |> String.split("\n")
  |> Stream.map(&calc.(&1))
  |> Enum.to_list()

IO.inspect(Enum.max(list))

list = Enum.sort(list)

Enum.reduce(list, fn i, acc ->
  if acc + 2 == i do
    IO.puts("found my seat #{acc + 1}")
  end

  i
end)

File: day05/main.go
package main

import (
	"bufio"
	"log"
	"os"
	"sort"
	"strings"
)

type intRange struct {
	min int
	max int
}

func NewIntRange(pattern string, start, end int) *intRange {
	r := new(intRange)
	r.min = start
	r.max = end
	for _, p := range strings.Split(pattern, "") {
		if p == "F" || p == "L" {
			r.lowerHalf()
		} else if p == "B" || p == "R" {
			r.upperHalf()
		}
	}
	return r
}

func (r *intRange) upperHalf() {
	tmpRange := r.max - r.min
	newRange := tmpRange / 2
	r.min = r.max - newRange
}

func (r *intRange) lowerHalf() {
	tmpRange := r.max - r.min
	newRange := tmpRange / 2
	r.max = r.min + newRange
}

func (r *intRange) result() int {
	return r.max
}

func main() {
	f, _ := os.Open("input.txt")
	scanner := bufio.NewScanner(f)
	scanner.Split(bufio.ScanLines)

	result := make([]int, 0)

	for scanner.Scan() {
		p := scanner.Text()
		row := NewIntRange(p[:7], 0, 127)
		seat := NewIntRange(p[7:], 0, 7)

		result = append(result, row.result()*8+seat.result())
	}

	sort.Ints(result)

	log.Println("part 1", result[len(result)-1])

	for i, p := range result {
		if p+2 == result[i+1] { // let it crash if nothing is found at the end of the list
			log.Println("part 2", p+1)
			return
		}
	}

}
